README.md
<span align="center">

# @switchbot/homebridge-switchbot-ble

[![npm version](https://badgen.net/npm/v/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)
[![npm downloads](https://badgen.net/npm/dt/@switchbot/homebridge-switchbot-ble)](https://www.npmjs.com/package/@switchbot/homebridge-switchbot-ble)

</span>

The @switchbot/homebridge-switchbot-ble is a nodejs module , and also a homebridge plug-in that directly controls SwitchBot products via BLE.

Now supports:

- [SwitchBot (Bot)](https://www.switch-bot.com/products/switchbot-bot)
- [SwitchBot Curtain (Curtain)](https://www.switch-bot.com/products/switchbot-curtain)
- [SwitchBot Thermometer & Hygrometer (Meter)](https://www.switch-bot.com/products/switchbot-meter)

## Installation

### Installing bluetooth libraries

##### Ubuntu, Debian, Raspbian

```sh
sudo apt-get install bluetooth bluez libbluetooth-dev libudev-dev
```

See the document of the [@abandonware/noble](https://github.com/abandonware/noble#readme) for other operating systems details.

### Installing package

```sh
sudo npm install -g @switchbot/homebridge-switchbot-ble
```

You can also install it on the homebridge plugins page.
![homebridge-plugins-search](image/homebridge-plugins-search.png)

## Configuration

Add this to your homebridge `config.json` file

```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform"
    }
]
```

If it work, add devices.

```json
"platforms": [
    {
        "platform": "SwitchBotPlatform",
        "name": "SwitchBotPlatform",
        "devices": [
            {
                "type": "bot",
                "name": "Bot 2c",
                "bleMac": "d4:cc:43:97:ae:2c"
            },
            {
                "type": "bot",
                "name": "Bot 51",
                "bleMac": "e7:4d:36:cf:9e:51",
                "scanDuration": 2000
            },
            {
                "type": "curtain",
                "name": "Curtain 11",
                "bleMac": "ec:58:c5:d0:01:11",
                "scanDuration": 2000,
                "reverseDir": false,
                "moveTime": 2000
            },
            {
                "type": "meter",
                "name": "Meter 84",
                "bleMac": "d0:d5:64:14:43:84",
                "scanDuration": 5000,
                "scanInterval": 60000
            }
        ]
    }
]
```

**Requird Settings**

- `devices` - SwitchBot devices list.
- `type` - Device type. Currently supports `bot`, `curtain` and `meter`.
- `name` - Device name.
- `bleMac` - Device mac address. You can find it in App settings.

**Optional Settings**

- `scanDuration` - Scan timeout. BLE Central must first scan the advertising. Default is `1000` (unit: ms). Longer time to ensure device discovery but slower response.
- `reverseDir` - Set to `true` to exchange the "opened" and "closed" directions of Curtain after calibration. Default is `false`. So you can swap the directions without recalibration.
- `moveTime` - Set the actual running time of Curtain, which is also the time to switch from running state to opened/closed state in the Home APP. Default is `2000` (unit: ms).
- `scanInterval`
  - For a Meter, sets how frequently to scan in order to retrieve temperature and humidity values. Default is `60000` (unit: ms).
  - For a Curtain, sets how frequently to scan in order to retrieve the curtain position. Default is `60000` (unit: ms) when idle.

Please note that:

1. The device `name` cannot be the same.
2. Restart HomeBridge after updating the config file.
3. The "opened" and "closed" directions of HomeKit are the same as those in Curtain and SwitchBot APP, but the percentage definitions are different. "Closed" is 0% in HomeKit but 100% in Curtain and APP.

## Community

- [SwitchBot (Official website)](https://www.switch-bot.com/)
- [Facebook @SwitchBotRobot](https://www.facebook.com/SwitchBotRobot/)
- [Twitter @SwitchBot](https://twitter.com/switchbot)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. This project uses [Semantic Versioning](https://semver.org/)

## [Version 1.3.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases/tag/v1.3.0) (2021-07-13)

### Changes

- Add retry support to address disconnects.
- House Keeping and Update dependencies.

## [Version 1.2.10](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-02-19)

### Changes

- House Keeping and Update dependencies.

## [Version 1.2.9](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-01-21)

### Changes

- Fixed issues from `v1.2.8`.

## [Version 1.2.8](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2021-01-21)

### Changes

- Update Dependencies
  - Node-SwitchBot v1.0.0
    - Fix for "No device was found" in MacOS

## [Version 1.2.7](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-22)

### Changes

- Fixed issues from `v1.2.6`.

## [Version 1.2.6](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update dependencies.

## [Version 1.2.5](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Fixed issue with `config.schema.json` not being packaged in `v1.2.4`.
- Update Dependencies.

## [Version 1.2.4](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-12-21)

### Changes

- Support Homebridge config setting UI. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/24))
- Fixed TypeError: this.config.devices is not iterable. (Thanks to [@donavanbecker](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/23))

## [Version 1.2.3](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-30)

### Changes

- Fixed the incorrect state of Bot in Home App when control fails.
- Modify log messages. (Thanks to [@dnicolson](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/15))

## [Version 1.2.2](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-19)

### Changes

- Fixed Curtain state will not be updated in the Home App.

## [Version 1.2.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-12)

### Changes

- Get temperature and humidity from Meter advertising periodically.

## [Version 1.2.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-11-06)

### Changes

- Fixed HomeKit control direction reversal issue.
- Add option for inverted curtain position. (Thanks to [@whatUwant](https://github.com/SwitchBot/homebridge-switchbot-ble/pull/4))

## [Version 1.1.0](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-29)

### Changes

- Add support for Curtain.

## [Version 1.0.1](https://github.com/OpenWonderLabs/homebridge-switchbot-ble/releases) (2020-10-22)

### Changes

- First public release, supports Bot.


src/bot-accessory.ts
/* eslint-disable brace-style */
/* eslint-disable prefer-const */
/* eslint-disable eqeqeq */
/* eslint-disable @typescript-eslint/no-unused-vars */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";
import { rejects } from "assert";

export class Bot implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly switchbot: any;

  private switchOn = false;
  private runTimer!: NodeJS.Timeout;

  // This property must be existent!!
  name: string;

  private readonly botService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    const SwitchBot = require("node-switchbot");
    this.switchbot = new SwitchBot();

    this.botService = new hap.Service.Switch(name);
    this.botService
      .getCharacteristic(hap.Characteristic.On)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Current state of Bot was returned: " + (this.switchOn ? "ON" : "OFF"));
        callback(undefined, this.switchOn);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        const targetState = value as boolean;
        clearTimeout(this.runTimer);
        if (targetState === this.switchOn) {
          log.info("Target state of Bot has not changed: " + (this.switchOn ? "ON" : "OFF"));
          this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
          callback();
        }
        // Target state has been changed.
        log.info("Target state of Bot setting: " + (targetState ? "ON" : "OFF"));
        this.switchbot
          .discover({ duration: this.scanDuration, model: "H", quick: true, id: this.bleMac })
          .then((device_list: any) => {
            log.info("Scan done.");
            let targetDevice: any = null;
            for (let device of device_list) {
              // log.info(device.modelName, device.address);
              if (device.address == this.bleMac) {
                targetDevice = device;
                break;
              }
            }
            if (!targetDevice) {
              log.info("No device was found during scan.");
              return new Promise((resolve, reject) => {
                reject(new Error("No device was found during scan."));
              });
            } else {
              log.info(targetDevice.modelName + " (" + targetDevice.address + ") was found.");
              // Set event handers
              targetDevice.onconnect = () => {
                // log.info('Connected.');
              };
              targetDevice.ondisconnect = () => {
                // log.info('Disconnected.');
              };
              log.info("Bot is running...");
              return this.setTargetDeviceState(targetDevice, targetState);
            }
          })
          .then(() => {
            log.info("Done.");
            this.switchOn = targetState;
            this.runTimer = setTimeout(() => {
              this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
            }, 500);
            log.info("Bot state has been set to: " + (this.switchOn ? "ON" : "OFF"));
            callback();
          })
          .catch((error: any) => {
            log.error(error);
            this.runTimer = setTimeout(() => {
              this.botService?.getCharacteristic(hap.Characteristic.On).updateValue(this.switchOn);
            }, 500);
            log.info("Bot state failed to be set to: " + (targetState ? "ON" : "OFF"));
            callback();
          });
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info("Bot '%s' created!", name);
  }

  async retry(max: number, fn: { (): any; (): Promise<any>; }): Promise<null> {
    return fn().catch( async (err: any) => {
      if (max == 0) {
        throw err;
      }
      this.log.info(err);
      this.log.info("Retrying");
      await this.switchbot.wait(1000);
      return this.retry(max - 1, fn);
    });
  }

  async setTargetDeviceState(targetDevice: any, targetState: boolean): Promise<null> {
    return await this.retry(5, () => {
      if (targetState) {
        return targetDevice.turnOn();
      } else {
        return targetDevice.turnOff();
      }
    });
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.botService];
  }
}


src/meter-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-inferrable-types */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes,
} from "homebridge";
import { rejects } from "assert";

export class Meter implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly scanInterval: number;

  private temperature: number = 0;
  private humidity: number = 0;

  // This property must be existent!!
  name: string;

  private readonly temperatureSercice: Service;
  private readonly humidityService: Service;
  private readonly informationService: Service;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.bleMac = bleMac;
    this.scanDuration = scanDuration;
    this.scanInterval = scanInterval;

    this.temperatureSercice = new hap.Service.TemperatureSensor(name);
    this.temperatureSercice
      .getCharacteristic(hap.Characteristic.CurrentTemperature)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current temperature: " + this.temperature + "\u2103");
        callback(undefined, this.temperature < 0 ? 0 : this.temperature > 100 ? 100 : this.temperature);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The temperature of the Meter can't be set!");
        callback();
      });

    this.humidityService = new hap.Service.HumiditySensor(name);
    this.humidityService
      .getCharacteristic(hap.Characteristic.CurrentRelativeHumidity)
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info(name + " current humidity: " + this.humidity + "%");
        callback(undefined, this.humidity);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        log.info("The humidity of the Meter can't be set!");
        callback();
      });

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-METERTH-S1")
      .setCharacteristic(hap.Characteristic.SerialNumber, this.bleMac);

    log.info(name, "scanDuration:" + this.scanDuration.toString() + "ms", "scanInterval:" + this.scanInterval.toString() + "ms");

    const Switchbot = require("node-switchbot");
    const switchbot = new Switchbot();

    switchbot.onadvertisement = (ad: any) => {
      // log.info(JSON.stringify(ad, null, '  '));
      // log.info("Temperature:", ad.serviceData.temperature.c);
      // log.info("Humidity:", ad.serviceData.humidity);
      this.temperature = ad.serviceData.temperature.c;
      this.humidity = ad.serviceData.humidity;
    };

    switchbot
      .startScan({
        id: bleMac,
      })
      .then(() => {
        return switchbot.wait(this.scanDuration);
      })
      .then(() => {
        switchbot.stopScan();
      })
      .catch((error: any) => {
        log.error(error);
      });

    setInterval(() => {
      // log.info("Start scan " + name + "(" + bleMac + ")");
      switchbot
        .startScan({
          // mode: 'T',
          id: bleMac,
        })
        .then(() => {
          return switchbot.wait(this.scanDuration);
        })
        .then(() => {
          switchbot.stopScan();
          // log.info("Stop scan " + name + "(" + bleMac + ")");
        })
        .catch((error: any) => {
          log.error(error);
        });
    }, this.scanInterval);
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.temperatureSercice, this.humidityService];
  }
}


src/curtain-accessory.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable eqeqeq */
/* eslint-disable prefer-const */
/* eslint-disable brace-style */
import {
  AccessoryPlugin,
  CharacteristicGetCallback,
  CharacteristicSetCallback,
  CharacteristicValue,
  HAP,
  Logging,
  Service,
  CharacteristicEventTypes, Characteristic,
} from "homebridge";

type OnPositionReceived = (pos: number) => void

class Device {
  private readonly bleMac: string;
  private readonly scanDuration: number;
  private readonly slowScanInterval: number;
  private readonly fastScanInterval = 500;
  private readonly fastScanDuration = 10000;

  private readonly name: string;
  private readonly log: Logging;
  private readonly callback: OnPositionReceived;
  private readonly considerClosedAtPosition = 98;

  private fastScanEnabled = false;
  private autoDisableFastScanTimeoutId: NodeJS.Timeout | null = null;

  private scanIntervalId: NodeJS.Timeout | null = null;

  private previousPosition = -1;
  private position = -1;

  constructor(bleMac: string, name: string, log: Logging, scanDuration: number, scanInterval: number, callback: OnPositionReceived) {
    this.name = name;
    this.bleMac = bleMac;
    this.log = log;

    this.scanDuration = scanDuration;
    this.slowScanInterval = scanInterval;
    this.callback = callback;

    this.startFastScan();
  }

  /**
   * Ask the device to start moving to the given position (which must be a device value and not an HomeKit value).
   * Returns a Promise that's resolved as soon as the command was sent to the device.
   */
  public runToPosition(pos: number): Promise<void> {
    const SwitchBot = require("node-switchbot");
    const switchbot = new SwitchBot();

    return switchbot
      .discover({duration: this.scanDuration, model: "c", quick: false})
      .then((device_list: any) => {
        let targetDevice: any = null;

        for (let device of device_list) {
          this.log.info(device.modelName, device.address);
          if (device.address == this.bleMac) {
            targetDevice = device;
            break;
          }
        }

        if (!targetDevice) {
          return new Promise((resolve, reject) => {
            reject(new Error("Curtain '" + this.name + "' (" + this.bleMac + "): device not found."));
          });
        }

        this.startFastScan();

        this.log.info("Curtain '%s' (%s) is moving to %d%%...", this.name, this.bleMac, pos);
        return targetDevice.runToPos(pos);
      });
  }

  /**
   * Start a faster scan loop (using fastScanInterval). Used in calls to `runToPosition` to
   * report quicker on device position change. Will disable on its own as soon if the curtain's
   * device position does not change for more than `fastScanDuration`.
   */
  private startFastScan() {
    if (this.fastScanEnabled) {
      return;
    }
    this.fastScanEnabled = true;
    this.startScanLoop();
  }

  private stopFastScan() {
    if (!this.fastScanEnabled) {
      return;
    }
    this.fastScanEnabled = false;
    this.startScanLoop();
  }

  private get scanInterval(): number {
    return this.fastScanEnabled ? this.fastScanInterval : this.slowScanInterval;
  }

  private startScanLoop() {
    if (this.scanIntervalId !== null) {
      clearInterval(this.scanIntervalId);
    }

    this.log.info("Curtain '%s': starting scan loop with interval %d", this.name, this.scanInterval);

    this.scanIntervalId = setInterval(() => {
      this.scan().catch((err) => {
        this.log.error("error while scanning for Curtain '%s': %s", this.name, err);
      });
    }, this.scanInterval);
  }

  private scan(): Promise<void> {
    return new Promise((resolve, reject) => {
      const SwitchBot = require("node-switchbot");
      const switchbot = new SwitchBot();
      switchbot.onadvertisement = (ad: any) => {
        this.applyPosition(ad);
      };
      switchbot.startScan({id: this.bleMac})
        .then(() => {
          return switchbot.wait(this.scanDuration);
        })
        .then(() => {
          resolve();
          switchbot.stopScan();
        })
        .catch((err: any) => {
          reject(err);
        });
    });
  }

  private applyPosition(ad: any) {
    let pos = ad.serviceData.position;

    if (pos >= this.considerClosedAtPosition) {
      pos = 100;
    }

    this.previousPosition = this.position;
    this.position = pos;

    if (this.position === this.previousPosition) {
      return;
    }

    if (this.autoDisableFastScanTimeoutId !== null) {
      clearTimeout(this.autoDisableFastScanTimeoutId);
    }

    this.autoDisableFastScanTimeoutId = setTimeout(() => {
      this.stopFastScan();
    }, this.fastScanDuration);

    this.callback(this.position);
  }
}

export class Curtain implements AccessoryPlugin {
  private readonly log: Logging;
  private readonly reverseDir: boolean;
  private readonly moveTime: number;

  private didReceiveInitialPosition = false;

  private currentPosition = 0;
  private targetPosition = 0;
  private positionState = 0;
  private moveTimer!: NodeJS.Timeout;

  private device: Device;

  // This property must be existent!!
  name: string;

  private readonly curtainService: Service;
  private readonly informationService: Service;

  private currentPositionCharacteristic: Characteristic;
  private targetPositionCharacteristic: Characteristic;
  private positionStateCharacteristic: Characteristic;

  constructor(hap: HAP, log: Logging, name: string, bleMac: string, scanDuration: number, reverseDir: boolean, moveTime: number, scanInterval: number) {
    this.log = log;
    this.name = name;
    this.reverseDir = reverseDir;
    this.moveTime = moveTime;

    this.device = new Device(bleMac, name, log, scanDuration, scanInterval, (pos: number) => {
      this.currentPosition = this.convertFromHomeKitPosition(pos);
      this.currentPositionCharacteristic.updateValue(this.currentPosition);

      // Set target to the same value as current on the first received position, or if the curtain is in
      // STOPPED state (eg. it's still or not moving from a change triggered by HomeKit). Otherwise, with
      // target = X and current != X, HomeKit will assume curtain is moving and show the progress indicator.
      //
      // Note that on an HomeKit-triggered change, `positionState` is set to STOPPED after `moveTime` has
      // passed, rather than when the curtain actually finished moving. When `moveTime` is lower than the
      // curtain's actual travel time, the device will appear in HomeKit first as moving (with a progress
      // indicator) then after `moveTime` has passed, as not moving (without progress indicator), but with
      // rapidly changing percentages.
      if (!this.didReceiveInitialPosition || this.positionState === hap.Characteristic.PositionState.STOPPED) {
        this.didReceiveInitialPosition = true;
        this.targetPosition = this.currentPosition;
        this.targetPositionCharacteristic.updateValue(this.targetPosition);
      }
    });

    this.positionState = hap.Characteristic.PositionState.STOPPED;

    this.informationService = new hap.Service.AccessoryInformation()
      .setCharacteristic(hap.Characteristic.Manufacturer, "SwitchBot")
      .setCharacteristic(hap.Characteristic.Model, "SWITCHBOT-CURTAIN-W0701600")
      .setCharacteristic(hap.Characteristic.SerialNumber, bleMac);

    this.curtainService = new hap.Service.WindowCovering(name);

    this.currentPositionCharacteristic = this.curtainService
      .getCharacteristic(hap.Characteristic.CurrentPosition);

    this.targetPositionCharacteristic = this.curtainService
      .getCharacteristic(hap.Characteristic.TargetPosition);

    this.positionStateCharacteristic = this.curtainService
      .getCharacteristic(hap.Characteristic.PositionState);

    this.currentPositionCharacteristic
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Current position of Curtain was returned: " + this.currentPosition + "%");
        callback(undefined, this.currentPosition);
      });

    this.targetPositionCharacteristic
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("Target position of Curtain was returned: " + this.targetPosition + "%");
        callback(undefined, this.targetPosition);
      })
      .on(CharacteristicEventTypes.SET, (value: CharacteristicValue, callback: CharacteristicSetCallback) => {
        this.targetPosition = value as number;
        log.info("Target position of Curtain setting: " + this.targetPosition + "%");
        clearTimeout(this.moveTimer);
        if (this.targetPosition > this.currentPosition) {
          this.positionState = hap.Characteristic.PositionState.INCREASING;
        } else if (this.targetPosition < this.currentPosition) {
          this.positionState = hap.Characteristic.PositionState.DECREASING;
        } else {
          this.positionState = hap.Characteristic.PositionState.STOPPED;
        }

        if (this.positionState === hap.Characteristic.PositionState.STOPPED) {
          this.targetPositionCharacteristic.updateValue(this.targetPosition);
          this.currentPositionCharacteristic.updateValue(this.currentPosition);
          this.positionStateCharacteristic.updateValue(this.positionState);
          callback();
        } else {
          this.device.runToPosition(this.convertFromHomeKitPosition(this.targetPosition))
            .then(() => {
              log.info("Done.");
              log.info("Target position of Curtain has been set to: " + this.targetPosition + "%");
              this.moveTimer = setTimeout(() => {
                // log.info("setTimeout", this.positionState.toString(), this.currentPosition.toString(), this.targetPosition.toString());
                this.positionState = hap.Characteristic.PositionState.STOPPED;
                // this.curtainService?.getCharacteristic(hap.Characteristic.TargetPosition).updateValue(this.targetPosition);
                this.currentPositionCharacteristic.updateValue(this.currentPosition);
                this.positionStateCharacteristic.updateValue(this.positionState);
              }, this.moveTime);
              callback();
            })
            .catch((error: any) => {
              log.error(error);
              this.moveTimer = setTimeout(() => {
                this.targetPosition = this.currentPosition;
                this.positionState = hap.Characteristic.PositionState.STOPPED;
                this.targetPositionCharacteristic.updateValue(this.targetPosition);
                // this.curtainService?.getCharacteristic(hap.Characteristic.CurrentPosition).updateValue(this.currentPosition);
                this.positionStateCharacteristic.updateValue(this.positionState);
              }, 1000);
              log.info("Target position of Curtain failed to be set to: " + this.targetPosition + "%");
              callback();
            });
        }
      });

    this.positionStateCharacteristic
      .on(CharacteristicEventTypes.GET, (callback: CharacteristicGetCallback) => {
        log.info("The position state of Curtain was returned: " + this.positionState);
        callback(undefined, this.positionState);
      });

    log.info("Curtain '%s' created!", name);
  }


  /**
   * Convert to/from device/HomeKit's position, since:
   *
   * - opened is 0% in HomeKit and 100% in Curtain device.
   * - closed is 100% in HomeKit, 0% in Curtain device.
   */
  convertFromHomeKitPosition(n: number): number {
    let covertToDevicePosition: number;
    if (this.reverseDir) {
      covertToDevicePosition = n;
    } else {
      covertToDevicePosition = 100 - n;
    }
    return covertToDevicePosition;
  }

  /*
   * This method is optional to implement. It is called when HomeKit ask to identify the accessory.
   * Typical this only ever happens at the pairing process.
   */
  identify(): void {
    this.log.info("Identify!");
  }

  /*
   * This method is called directly after creation of this instance.
   * It should return all services which should be added to the accessory.
   */
  getServices(): Service[] {
    return [this.informationService, this.curtainService];
  }
}


src/switchbot-platform.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-var */
/* eslint-disable no-case-declarations */
/* eslint-disable prefer-const */
import { AccessoryPlugin, API, HAP, Logging, PlatformConfig, StaticPlatformPlugin } from "homebridge";
import { Bot } from "./bot-accessory";
import { Curtain } from "./curtain-accessory";
import { Meter } from "./meter-accessory";

const PLATFORM_NAME = "SwitchBotPlatform";

/*
 * IMPORTANT NOTICE
 *
 * One thing you need to take care of is, that you never ever ever import anything directly from the "homebridge" module (or the "hap-nodejs" module).
 * The above import block may seem like, that we do exactly that, but actually those imports are only used for types and interfaces
 * and will disappear once the code is compiled to Javascript.
 * In fact you can check that by running `npm run build` and opening the compiled Javascript file in the `dist` folder.
 * You will notice that the file does not contain a `... = require("homebridge");` statement anywhere in the code.
 *
 * The contents of the above import statement MUST ONLY be used for type annotation or accessing things like CONST ENUMS,
 * which is a special case as they get replaced by the actual value and do not remain as a reference in the compiled code.
 * Meaning normal enums are bad, const enums can be used.
 *
 * You MUST NOT import anything else which remains as a reference in the code, as this will result in
 * a `... = require("homebridge");` to be compiled into the final Javascript code.
 * This typically leads to unexpected behavior at runtime, as in many cases it won't be able to find the module
 * or will import another instance of homebridge causing collisions.
 *
 * To mitigate this the {@link API | Homebridge API} exposes the whole suite of HAP-NodeJS inside the `hap` property
 * of the api object, which can be acquired for example in the initializer function. This reference can be stored
 * like this for example and used to access all exported variables and classes from HAP-NodeJS.
 */
let hap: HAP;

export = (api: API) => {
  hap = api.hap;

  api.registerPlatform(PLATFORM_NAME, SwitchBotPlatform);
};

class SwitchBotPlatform implements StaticPlatformPlugin {
  private readonly log: Logging;
  private readonly config: PlatformConfig;

  constructor(log: Logging, config: PlatformConfig, api: API) {
    this.log = log;

    // probably parse config or something here
    this.config = config;

    log.info("SwitchBot platform finished initializing!");
  }

  /*
   * This method is called to retrieve all accessories exposed by the platform.
   * The Platform can delay the response my invoking the callback at a later time,
   * it will delay the bridge startup though, so keep it to a minimum.
   * The set of exposed accessories CANNOT change over the lifetime of the plugin!
   */
  accessories(callback: (foundAccessories: AccessoryPlugin[]) => void): void {
    let deviceList = [];
    if (this.config.devices) {
      for (var device of this.config.devices) {
        // this.log.info(device.type);
        // this.log.info(device.name);
        // this.log.info(device.bleMac);
        // this.log.info(device.scanDuration, typeof device.scanDuration);
        let scanDuration: number = device.scanDuration || 1000;
        switch (device.type) {
          case "bot":
            deviceList.push(new Bot(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration));
            break;
          case "curtain":
            const reverseDir: boolean = device.reverseDir || false;
            const moveTime: number = device.moveTime || 2000;
            deviceList.push(new Curtain(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, reverseDir, moveTime, device.scanInterval || 60000));
            break;
          case "meter":
            let scanInterval: number = device.scanInterval || 60000;
            if (scanInterval < scanDuration) {
              scanInterval = scanDuration + 1000;
            }
            deviceList.push(new Meter(hap, this.log, device.name, device.bleMac.toLowerCase(), scanDuration, scanInterval));
            break;
          default:
            break;
        }
      }
    } else {
      this.log.error("No Device Set In Config");
    }
    this.log("Device amount:", deviceList.length.toString());
    callback(deviceList);
  }
}
